import/dproto/intermediate.d

```diff
@@ -267,3 +291,69 @@ struct Field {
 		sink("break;\n");
 	}
 }
+
+
+struct Service {
+	this(string name) {
+		this.name = name.idup;
+	}
+	string name;
+	Options options;
+	Method[] rpc;
+
+	struct Method {
+		this(string name, string documentation, string requestType, string responseType) {
+			this.name = name.idup;
+			this.documentation = documentation.idup;
+			this.requestType = requestType.idup;
+			this.responseType = responseType.idup;
+		}
+		string name;
+		string documentation;
+		string requestType;
+		string responseType;
+		Options options;
+
+		const void toString(scope void delegate(const(char)[]) sink, FormatSpec!char fmt)
+		{
+			switch(fmt.spec) {
+				case 'p':
+					sink.formattedWrite("rpc %s (%s) returns (%s)", name, requestType, responseType);
+					if(options.length > 0) {
+						sink(" {\n");
+						foreach(opt, val; options) {
+							sink.formattedWrite("option %s = %s;\n", opt, val);
+						}
+						sink("}\n");
+					} else {
+						sink(";\n");
+					}
+					break;
+				case 'i':
+					sink.formattedWrite("%s %s (%s); { %s res; return res; }\n", responseType, name, requestType);
+					break;
+				default:
+					sink.formattedWrite("%s %s (%s);\n", responseType, name, requestType);
+					break;
+			}
+		}
+	}
+
+	const void toString(scope void delegate(const(char)[]) sink, FormatSpec!char fmt)
+	{
+		switch(fmt.spec) {
+			case 'p':
+				sink.formattedWrite("service %s {\n", name);
+				break;
+			case 'i':
```

Comment:
%i - "implemented class". At present, this creates a class with stub
methods, for quick testing/rapid prototyping.
It's not as graceful as I'd like... I'd rather pass flags, but it's hacky
to use precision/width on settings like this.

Do you have some suggestions on how we might send flags?
FormatSpec seems to hold a single char, as well as width/precision
arguments.

It seemed like we could use those to specify additional options to the
toString methods.
For instance: %3p could mean "output protobuf version 3".

Alternatively, if we keep a context variable around, all pointing to a
parent class, then the toString methods
could simply access that for flag info (though this makes them more tightly
coupled).

I'm going to have a similar issue with JSON. It's just a minor tweak (much
like implemented class), but the flags for it would be weird.

-Charles



On Thu, Mar 19, 2015 at 7:48 AM, Matt Soucy <notifications@github.com>
wrote:

> In import/dproto/intermediate.d
> <https://github.com/msoucy/dproto/pull/40#discussion_r26760384>:
>
> > +					sink.formattedWrite("%s %s (%s); { %s res; return res; }\n", responseType, name, requestType);
> > +					break;
> > +				default:
> > +					sink.formattedWrite("%s %s (%s);\n", responseType, name, requestType);
> > +					break;
> > +			}
> > +		}
> > +	}
> > +
> > +	const void toString(scope void delegate(const(char)[]) sink, FormatSpec!char fmt)
> > +	{
> > +		switch(fmt.spec) {
> > +			case 'p':
> > +				sink.formattedWrite("service %s {\n", name);
> > +				break;
> > +			case 'i':
>
> "%i" meaning what? The names that I had in place are "%d" -> D, "%p" ->
> proto
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/msoucy/dproto/pull/40/files#r26760384>.
>